cleanup.sh

```diff
@@ -0,0 +1,63 @@
+#!/usr/bin/env bash
+
+# This script does not need to be changed for certbots DNS challenge.
+# Please see the .dedynauth file for authentication information.
+
+(
+
+DEDYNAUTH=$(pwd)/.dedynauth
+
+if [ ! -f .dedynauth ]; then
+    >&2 echo "File $DEDYNAUTH not found. Please place .dedynauth file in appropriate location."
+    exit 1
+fi
+
+source .dedynauth
+
+if [ -z "$DEDYN_TOKEN" ]; then
+    >&2 echo "Variable \$DEDYN_TOKEN not found. Please set DEDYN_TOKEN=(your dedyn.io token) to your dedyn.io access token in $DEDYNAUTH, e.g."
+    >&2 echo ""
+    >&2 echo "DEDYN_TOKEN=d41d8cd98f00b204e9800998ecf8427e"
+    exit 2
+fi
+
+if [ -z "$DEDYN_NAME" ]; then
+    >&2 echo "Variable \$DEDYN_NAME not found. Please set DEDYN_NAME=(your dedyn.io name) to your dedyn.io name in $DEDYNAUTH, e.g."
+    >&2 echo ""
+    >&2 echo "DEDYN_NAME=foobar.dedyn.io"
+    exit 3
+fi
+
+if [ -z "$CERTBOT_DOMAIN" ]; then
+    >&2 echo "It appears that you are not running this script through certbot (\$CERTBOT_DOMAIN is unset). Please call with: certbot --manual-auth-hook=$0"
+    exit 4
+fi
+
+if [ ! "$(type -P curl)" ]; then
+    >&2 echo "Please install curl to use certbot with dedyn.io."
+    exit 5
+fi
+
+echo "Deleting challenge(s) for ${CERTBOT_DOMAIN} ..."
+
+# Figure out subdomain infix by removing zone name and trailing dot
+# foobar.dedyn.io gives "" while a.foobar.dedyn.io gives ".a"
+domain=.$CERTBOT_DOMAIN
+infix=${domain%.$DEDYN_NAME}
+
+# Remove leading wildcard from infix, if present
+# *.foobar.dedyn.io gives "" while *.a.foobar.dedyn.io gives ".a"
+infix=${infix#.\*}
+
+args=( \
+    '-sSLf' \
+    '-o' '/dev/null' \
+    '-H' "Authorization: Token $DEDYN_TOKEN" \
+)
+
+# Delete ACME challenge(s)
+curl "${args[@]}" -X DELETE "https://desec.io/api/v1/domains/$DEDYN_NAME/rrsets/_acme-challenge$infix/TXT/"
```

Comment:
Removing only the current challenge does seem cleaner to me, but without relying on `jq` we're looking at a lot of regex... I'll play with it and see what I break.

Reply:
Awesome idea!

I suggest the following, which requires neither `@` nor the above `if`:

    acme_records=$acme_records,  # adds comma
    acme_records=${acme_records//*([[:space:]])'"\"'"$CERTBOT_VALIDATION"'\""'*([[:space:]])','*([[:space:]])/}  # removes all occurrences (double-slash)
    acme_records=${acme_records:0:-1}  # strip final comma (only if non-zero length)

The result can then be directly used in the `PUT` payload, like `"records": [$acme_records]`.

The `*([[:space:]])` patterns are there so that the matching does not depend on whitespace at the beginning of the string or before/after each comma. Because of this, `shopt -s extglob` is still required.

What do you think?

Reply:
How are you figuring "only if non-zero length"? For me, I get:
```
"\"test1\""
"\"test1\"",

-bash: -1: substring expression < 0
```

Reply:
I meant that a test would be required, yes

Reply:
Works perfect!